Spigot-API-Patches/0157-Turtle-API.patch

```diff
@@ -0,0 +1,278 @@
++
++import org.bukkit.entity.Slime;
++import org.bukkit.entity.Turtle;
++import org.bukkit.event.Cancellable;
++import org.bukkit.event.HandlerList;
++import org.bukkit.event.entity.EntityEvent;
++
++/**
++ * Fired when a Turtle decides to go home
++ */
++public class TurtleGoHomeEvent extends EntityEvent implements Cancellable {
++    private static final HandlerList handlers = new HandlerList();
++    private boolean cancelled = false;
++
++    public TurtleGoHomeEvent(Turtle turtle) {
++        super(turtle);
++    }
++
++    /**
++     * The turtle going home
++     *
++     * @return The turtle
++     */
++    public Turtle getEntity() {
++        return (Turtle) entity;
++    }
++
++    @Override
++    public boolean isCancelled() {
++        return cancelled;
++    }
++
++    @Override
++    public void setCancelled(boolean cancel) {
++        cancelled = cancel;
++    }
++
++    public HandlerList getHandlers() {
++        return handlers;
++    }
++
++    public static HandlerList getHandlerList() {
++        return handlers;
++    }
++}
+diff --git a/src/main/java/com/destroystokyo/paper/event/entity/TurtleLayEggEvent.java b/src/main/java/com/destroystokyo/paper/event/entity/TurtleLayEggEvent.java
+new file mode 100644
+index 00000000..a2f0dc92
+--- /dev/null
++++ b/src/main/java/com/destroystokyo/paper/event/entity/TurtleLayEggEvent.java
+@@ -0,0 +1,81 @@
++package com.destroystokyo.paper.event.entity;
++
++import org.bukkit.Location;
++import org.bukkit.entity.Turtle;
++import org.bukkit.event.Cancellable;
++import org.bukkit.event.HandlerList;
++import org.bukkit.event.entity.EntityEvent;
++
++/**
++ * Fired when a Turtle lays eggs
++ */
++public class TurtleLayEggEvent extends EntityEvent implements Cancellable {
++    private static final HandlerList handlers = new HandlerList();
++    private boolean cancelled = false;
++    private final Location location;
++    private int eggCount;
++
++    public TurtleLayEggEvent(Turtle turtle, Location location, int eggCount) {
++        super(turtle);
++        this.location = location;
++        this.eggCount = eggCount;
++    }
++
++    /**
++     * The turtle going home
++     *
++     * @return The turtle
++     */
++    public Turtle getEntity() {
++        return (Turtle) entity;
++    }
++
++    /**
++     * Get the location where the eggs are being laid
++     *
++     * @return Location of eggs
++     */
++    public Location getLocation() {
++        return location;
++    }
++
++    /**
++     * Get the number of eggs being laid
++     *
++     * @return Number of eggs
++     */
++    public int getEggCount() {
++        return eggCount;
++    }
++
++    /**
++     * Set the number of eggs being laid
++     *
++     * @param eggCount Number of eggs
++     */
++    public void setEggCount(int eggCount) {
++        if (eggCount < 1) {
++            cancelled = true;
++            return;
++        }
++        eggCount = Math.min(eggCount, 4);
```

Comment:
In my opinion this maximum should be mentioned in the documentation.

Reply:
I was told before not to document implementation details.

Reply:
Yeah, implementation details should not be documented, but valid inputs should be. If I call this method with 5, I may wonder otherwise why it has no effect at all. I'm open to other opinions regarding this matter, but from the perspective of a user of this API one may wonder why his input is rectified.

And the valid input documentation vs implementation detail documentation thing could be explained by these examples:

Implementation detail:
```java
/**
  * Through the use of {@link Math#min(Integer, Integer)} the value is clamped to 4.
  **/
```

Valid Input:
```java
/**
  * <description>
  *
  * @param eggCount Number of eggs. This value is clamped between 0 (cancellation of event) and 4. Overflowing values will be rectified.
  **/
```

Reply:
If it's a restriction we enforce on our own, should be documented, but i'm iffy on documenting every vanilla restriction, as vanilla restrictions can change.

I don't see it as bad if the documentation said "As of MC 1.13.2, the max value is camped to 4" though.